stdlib/public/TensorFlow/TensorFlow+NumPy.swift

```diff
@@ -0,0 +1,86 @@
+//===-- TensorFlow+NumPy.swift --------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines bridging from `numpy.ndarray` to `ShapedArray` and
+// `Tensor`.
+//
+//===----------------------------------------------------------------------===//
+
+import Python
```

Comment:
I think we should add a `#if canImport(Python)` check so it would be possible to build the TensorFlow module without Python support with the minimum changes required.

Reply:
`#if canImport(Python)` (like other `#if` directives) is a static check, not a runtime one.
Thus, build order is important and a dependency is unavoidable.
If `TensorFlow` didn't depend on `Python` (or vice versa), then `#if canImport(...)` would evaluate to false and the code within wouldn't be compiled.

Reply:
I know that. But if somebody wants to build the TensorFlow module without the Python module that should be possible without removing source files. Adding this check allows that.

For example, if a `SWIFT_ENABLE_PYTHON_INTEROP` is added this check will ensure that the TensorFlow module remains buildable without removing the source file from the build.

Reply:
> I know that. But if somebody wants to build the TensorFlow module without the Python module that should be possible without removing source files. Adding this check allows that.

This is certainly ideal. It might even be possible with the current build setup, because different invocations of [`add_swift_library`](https://github.com/google/swift/blob/c4e3f745d55d272b9b49d7e1b4e05083609a4dc6/cmake/modules/AddSwift.cmake#L1403) aren't parallelized so a natural build order is produced based on invocations of [`add_subdirectory`](https://github.com/google/swift/blob/e2d98450e8906c1ddd58b498d846673dfb871e92/stdlib/public/CMakeLists.txt#L59).

But really, both libraries *are* dependencies for the code currently within `#if canImport(Python)`.
In theory, the code could be pulled into a separate Swift library like `TensorFlowPython` which depends on `TensorFlow` and `Python` (so that `TensorFlow` and `Python` don't have a dependency) but this is bad because it requires an extra import for users.

> For example, if a `SWIFT_ENABLE_PYTHON_INTEROP` is added this check will ensure that the TensorFlow module remains buildable without removing the source file from the build.

It's important to add a CMake flag like `SWIFT_ENABLE_PYTHON_INTEROP`, but I don't see how this solves the dependency problem.

Reply:
> It's important to add a CMake flag like SWIFT_ENABLE_PYTHON_INTEROP, but I don't see how this solves the dependency problem.

It doesn't. The idea would be that if you enable TensorFlow and Python on the build, the TensorFlow module would depend on Python but otherwise it should not.

Reply:
> It doesn't. The idea would be that if you enable TensorFlow and Python on the build, the TensorFlow module would depend on Python but otherwise it should not.

I see, that would be ideal.
Adding such a conditional dependency in CMake would probably require adding ad-hoc rules because currently there's only [`SWIFT_MODULE_DEPENDS`](https://github.com/google/swift/pull/31/commits/d8dfa71504fd1619f3c749fcec2022d7064fe90c#diff-93de774bf9b65f2e59e9adb120ec161eR61), which is unconditional.

Reply:
Yes, exactly!

Reply:
Can we add a `build-script` flag?
```bash
build-script --enable-tensorflow --tensorflow-with-python-extensions
```

Reply:
Yes. Adding a flag is the easy part, the difficult part is implementing the conditional dependency.

A straightforward implementation is to dupe the logic for handling `SWIFT_MODULE_DEPENDS` in `add_swift_library` (probably calls `add_dependencies` among other things) and paste it ad-hoc in `TensorFlow/CMakeLists.txt`.

Reply:
In any case, it is completely reasonable to make TensorFlow depend on Python for now.